### PR TITLE
cluster_name var  to change the name of cluster

### LINKED ISSUE #458: Create a dynamique name of cluster
This issue is to create a dynamique name of cluster

### PR DESCRIPTION
**Pull request by taabich**
#### Which issue(s) this PR fixes:

Fixes #458 

#### Additional comments:

The variable cluster_name ca be used to override the default name "mycluster" of the cluster

To make clear that you license your contribution under
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0) and that you give permission to [TOSIT](https://www.tosit.io/),
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [x] I hereby agree to grant [TOSIT](https://www.tosit.io/) a copyright license to use my contributions.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -107,6 +107,11 @@ Example usage:
 - debug:
     msg: "{{ groups['hdfs_jn'] | map('access_fqdn', hostvars) | list }}"
 ```
+## Configuration cluster
+
+| Variable | Description |
+| ------------------------------| --------------------------------------------------|
+| `cluster_name` | The default cluster name `mycluster`. |
```
Comment (leopaul36) [inline README.md:114]: No need to describe vars in the README for now

File: docs/developer.md [hunk 2]
```diff
@@ -84,8 +84,8 @@ By default, given the below hash:
 ```yaml
 hdfs_site:
   dfs.replication: 1
-  dfs.nameservices: mycluster
-  dfs.ha.namenodes.mycluster: nn1,nn2
+  dfs.nameservices: {{cluster_name | default("cluster_name")}}
```
Comment (leopaul36) [inline docs/developer.md:87]: Please respect our current conventions by keeping a whitespace before and after every `{{` and `}}`

File: roles/hdfs/ranger/tasks/init.yml [hunk 3]
```diff
@@ -14,7 +14,7 @@
         username: hdfs
         password: hdfs
         policy.download.auth.users: hdfs
-        fs.default.name: hdfs://mycluster
+        fs.default.name: hdfs://{{cluster_name |default("cluster_name")}}
```
Comment (leopaul36) [inline roles/hdfs/ranger/tasks/init.yml:17]: Please respect our current conventions by keeping a whitespace before and after every `{{` and `}}`

File: docs/developer.md [hunk 4]
```diff
@@ -84,8 +84,8 @@ By default, given the below hash:
...
+  dfs.ha.namenodes.{{cluster_name |default("cluster_name")}}: nn1,nn2
```
Comment (leopaul36) [inline docs/developer.md:88]: Let's not replace the vars in this static documentation

File: docs/developer.md [hunk 5]
```diff
@@ -95,7 +95,7 @@ hdfs_site:
   dfs.replication: 1
 ```
 
-... will replace `hdfs_site` completely, removing `dfs.nameservices` and `dfs.ha.namenodes.mycluster` from the `hdfs_site` hash.
+... will replace `hdfs_site` completely, removing `dfs.nameservices` and `dfs.ha.namenodes.{{cluster_name}}` from the `hdfs_site` hash.
```
Comment (leopaul36) [inline docs/developer.md:98]: Let's not replace the vars in this static documentation

File: tdp_vars_defaults/hadoop/hadoop.yml [hunk 6]
```diff
@@ -111,7 +111,7 @@ hadoop_ha_zookeeper_quorum: |
 
 # core-site.xml - common
 core_site:
-  fs.defaultFS: "hdfs://mycluster"
+  fs.defaultFS: "hdfs://{{cluster_name |default('mycluster')}}"
```
Comment (leopaul36) [inline tdp_vars_defaults/hadoop/hadoop.yml:114]: Please respect our current conventions by keeping a whitespace before and after every `{{` and `}}`

File: tdp_vars_defaults/hbase/hbase.yml [hunk 7]
```diff
@@ -79,7 +79,7 @@ hbase_zookeeper_client_port: "2181"
 # hbase_site.xml - common
 # TODO: make a hbase_site per service: master, rs, rest, client, phoenix queryserver
 hbase_site:
-  hbase.rootdir: hdfs://mycluster/hbase
+  hbase.rootdir: hdfs://{{cluster_name |default('mycluster') }}/hbase
   hbase.cluster.distributed: "true"
```
Comment (leopaul36) [inline tdp_vars_defaults/hbase/hbase.yml:83]: Please respect our current conventions by keeping a whitespace before and after every `{{` and `}}`

File: tdp_vars_defaults/hbase/hbase.yml [hunk 8]
```diff
@@ -123,7 +123,7 @@ ranger_hbase_install_properties:
   REPOSITORY_NAME: hbase-tdp
   XAAUDIT_SOLR_ENABLE: "{% if 'ranger_solr' in groups and groups['ranger_solr'] %}true{% else %}false{% endif %}"
   XAAUDIT_SOLR_URL: "{% if 'ranger_solr' in groups and groups['ranger_solr'] %}http://{{ groups['ranger_solr'][0] | tosit.tdp.access_fqdn(hostvars) }}:8983/solr/ranger_audits{% else %}NONE{% endif %}"
-  audit_hdfs: hdfs://mycluster
+  audit_hdfs: hdfs://{{cluster_name |default('mycluster')}}
```
Comment (leopaul36) [inline tdp_vars_defaults/hbase/hbase.yml:126]: Please respect our current conventions by keeping a whitespace before and after every `{{` and `}}`

File: tdp_vars_defaults/hdfs/hdfs.yml [hunk 9]
```diff
@@ -11,36 +11,36 @@ hadoop_hdfs_log_dir: /var/log/hadoop/hdfs
 # hdfs-site.xml - common
 # TODO: make a hdfs_site per service: nn, jn, dn
 hdfs_site:
-  dfs.nameservices: mycluster
-  dfs.ha.namenodes.mycluster: nn1,nn2
+  dfs.nameservices: "{{ cluster_name |default('myclyster') }}"
+  dfs.ha.namenodes.{{cluster_name |default('myclyster')}}: nn1,nn2
```
Comment (leopaul36) [inline tdp_vars_defaults/hdfs/hdfs.yml:15]: Please respect our current conventions by keeping a whitespace before and after every `{{` and `}}`

File: tdp_vars_defaults/hdfs/hdfs.yml [hunk 10]
```diff
@@ -11,36 +11,36 @@ hadoop_hdfs_log_dir: /var/log/hadoop/hdfs
...
-  dfs.namenode.http-address.mycluster.nn2: "{{ groups['hdfs_nn'][1] | tosit.tdp.access_fqdn(hostvars) }}:9870"
-  dfs.namenode.https-address.mycluster.nn1: "{{ groups['hdfs_nn'][0] | tosit.tdp.access_fqdn(hostvars) }}:9871"
-  dfs.namenode.https-address.mycluster.nn2: "{{ groups['hdfs_nn'][1] | tosit.tdp.access_fqdn(hostvars) }}:9871"
+  dfs.namenode.rpc-address.{{cluster_name |default('mycluster') }}.nn1: "{{ groups['hdfs_nn'][0] | tosit.tdp.access_fqdn(hostvars) }}:8020"
+  dfs.namenode.rpc-address.{{cluster_name |default('mycluster') }}.nn2: "{{ groups['hdfs_nn'][1] | tosit.tdp.access_fqdn(hostvars) }}:8020"
+  dfs.namenode.http-address.{{cluster_name |default('mycluster') }}.nn1: "{{ groups['hdfs_nn'][0] | tosit.tdp.access_fqdn(hostvars) }}:9870"
+  dfs.namenode.http-address.{{cluster_name |default('mycluster') }}.nn2: "{{ groups['hdfs_nn'][1] | tosit.tdp.access_fqdn(hostvars) }}:9870"
+  dfs.namenode.https-address.{{cluster_name |default('mycluster') }}.nn1: "{{ groups['hdfs_nn'][0] | tosit.tdp.access_fqdn(hostvars) }}:9871"
+  dfs.namenode.https-address.{{cluster_name |default('mycluster')}}.nn2: "{{ groups['hdfs_nn'][1] | tosit.tdp.access_fqdn(hostvars) }}:9871"
```
Comment (leopaul36) [inline tdp_vars_defaults/hdfs/hdfs.yml:28]: Please respect our current conventions by keeping a whitespace before and after every `{{` and `}}`

File: tdp_vars_defaults/hdfs/hdfs.yml [hunk 11]
```diff
@@ -11,36 +11,36 @@ hadoop_hdfs_log_dir: /var/log/hadoop/hdfs
...
   dfs.namenode.name.dir: "{{ hadoop_hdfs_dir }}/nn"
   dfs.namenode.shared.edits.dir: |-
     qjournal://{{ groups['hdfs_jn'] |
        map('tosit.tdp.access_fqdn', hostvars) |
        map('regex_replace', '^(.*)$', '\1:8485') |
        list |
        join(';')
-    }}/mycluster
+    }}/{{cluster_name |default('mycluster') }}
```
Comment (leopaul36) [inline tdp_vars_defaults/hdfs/hdfs.yml:36]: Please respect our current conventions by keeping a whitespace before and after every `{{` and `}}`

File: tdp_vars_defaults/hdfs/hdfs.yml [hunk 12]
```diff
@@ -11,36 +11,36 @@ hadoop_hdfs_log_dir: /var/log/hadoop/hdfs
...
 
   dfs.namenode.kerberos.principal: "nn/_HOST@{{ realm }}"
   dfs.namenode.keytab.file: /etc/security/keytabs/nn.service.keytab
   dfs.namenode.kerberos.internal.spnego.principal: "HTTP/_HOST@{{ realm }}"
   dfs.namenode.num.checkpoints.retained: 3
   dfs.namenode.name.dir.restore: true
-  dfs.client.failover.proxy.provider.mycluster: org.apache.hadoop.hdfs.server.namenode.ha.ConfiguredFailoverProxyProvider
+  dfs.client.failover.proxy.provider.{{cluster_name |default('mycluster') }}: org.apache.hadoop.hdfs.server.namenode.ha.ConfiguredFailoverProxyProvider
```
Comment (leopaul36) [inline tdp_vars_defaults/hdfs/hdfs.yml:43]: Please respect our current conventions by keeping a whitespace before and after every `{{` and `}}`

File: tdp_vars_defaults/hive/hive.yml [hunk 13]
```diff
@@ -118,7 +118,7 @@ ranger_hive_release: ranger-2.0.1-TDP-0.1.0-SNAPSHOT-hive-plugin
 ranger_hive_dist_file: "{{ ranger_hive_release }}.tar.gz"
 ranger_hive_install_dir: "{{ hive_root_dir }}/ranger-hive-plugin"
 ranger_hive_install_properties:
-  audit_hdfs: hdfs://mycluster
+  audit_hdfs: hdfs://{{cluster_name |default('mycluster') }}
```
Comment (leopaul36) [inline tdp_vars_defaults/hive/hive.yml:121]: Please respect our current conventions by keeping a whitespace before and after every `{{` and `}}`

File: tdp_vars_defaults/knox/knox.yml [hunk 14]
```diff
@@ -154,7 +154,7 @@ ranger_knox_release: ranger-2.0.1-TDP-0.1.0-SNAPSHOT-knox-plugin
 ranger_knox_dist_file: "{{ ranger_knox_release }}.tar.gz"
 ranger_knox_install_dir: "{{ knox_root_dir }}/ranger-knox-plugin"
 ranger_knox_install_properties:
-  audit_hdfs: hdfs://mycluster
+  audit_hdfs: hdfs://{{cluster_name |default('mycluster')}}
```
Comment (leopaul36) [inline tdp_vars_defaults/knox/knox.yml:157]: Please respect our current conventions by keeping a whitespace before and after every `{{` and `}}`

File: tdp_vars_defaults/spark/spark.yml [hunk 15]
```diff
@@ -77,7 +77,7 @@ spark_defaults_common:
 
 # spark-default.conf - Spark Client
 spark_defaults_client:
-  spark.eventLog.dir: hdfs://mycluster/spark-logs
+  spark.eventLog.dir: hdfs://{{cluster_name |default('mycluster') }}/spark-logs
```
Comment (leopaul36) [inline tdp_vars_defaults/spark/spark.yml:80]: Please respect our current conventions by keeping a whitespace before and after every `{{` and `}}`

File: tdp_vars_defaults/spark/spark.yml [hunk 16]
```diff
@@ -112,7 +112,7 @@ spark_defaults_hs:
   spark.ssl.historyServer.keyStore: "{{ spark_keystore_location }}"
   spark.ssl.historyServer.keyStorePassword: "{{ spark_keystore_password }}"
   spark.ssl.historyServer.port: 18081
-  spark.history.fs.logDirectory: hdfs://mycluster/spark-logs
+  spark.history.fs.logDirectory: hdfs://{{cluster_name |default('mycluster')}}/spark-logs
```
Comment (leopaul36) [inline tdp_vars_defaults/spark/spark.yml:115]: Please respect our current conventions by keeping a whitespace before and after every `{{` and `}}`

File: tdp_vars_defaults/spark3/spark3.yml [hunk 17]
```diff
@@ -72,7 +72,7 @@ spark_defaults_common:
 
 # spark-defaults.conf - Spark Client
 spark_defaults_client:
-  spark.eventLog.dir: hdfs://mycluster/spark3-logs
+  spark.eventLog.dir: hdfs://{{cluster_name|default('mycluster')}}/spark3-logs
```
Comment (leopaul36) [inline tdp_vars_defaults/spark3/spark3.yml:75]: Please respect our current conventions by keeping a whitespace before and after every `{{` and `}}`

File: tdp_vars_defaults/spark3/spark3.yml [hunk 18]
```diff
@@ -108,7 +108,7 @@ spark_defaults_hs:
   spark.ssl.historyServer.keyStore: "{{ spark_keystore_location }}"
   spark.ssl.historyServer.keyStorePassword: "{{ spark_keystore_password }}"
   spark.ssl.historyServer.port: 18083
-  spark.history.fs.logDirectory: hdfs://mycluster/spark3-logs
+  spark.history.fs.logDirectory: hdfs://{{cluster_name |default('mycluster')}}/spark3-logs
```
Comment (leopaul36) [inline tdp_vars_defaults/spark3/spark3.yml:111]: Please respect our current conventions by keeping a whitespace before and after every `{{` and `}}`

File: tdp_vars_defaults/yarn/yarn.yml [hunk 19]
```diff
@@ -25,7 +25,7 @@ yarn_site:
   yarn.nodemanager.remote-app-log-dir-suffix : "logs"
   yarn.resourcemanager.ha.enabled: "true"
   yarn.resourcemanager.ha.rm-ids: "rm1,rm2"
-  yarn.resourcemanager.cluster-id: "mycluster"
+  yarn.resourcemanager.cluster-id: "{{cluster_name |default('mycluster')}}"
```
Comment (leopaul36) [inline tdp_vars_defaults/yarn/yarn.yml:28]: Please respect our current conventions by keeping a whitespace before and after every `{{` and `}}`

**Review by leopaul36 — CHANGES_REQUESTED** [review]
Hello @taabich,

Thanks for your contribution to TDP.

Additionally to my comments, please recommit using [SemVer](https://semver.org/) for commit messages as stated in our [contributing](https://github.com/TOSIT-IO/TDP/blob/main/CONTRIBUTING.md#semantic-versioning) docs.

**Comment by taabich** [comment]
Hi,
I committed the conventions to respect space before {{ and after }}.
The role of template-plus is to evaluate the variables in the keys of ansible dictionary: Standard templates  can't do it.
Exemple: 
Template Ansible can't evalute fs.namenode.rpc-address.{{ cluster_name | default('mycluster') }}.nn1 but template-plus do it.
hdfs_site:
...
dfs.namenode.rpc-address.{{ cluster_name | default('mycluster') }}.nn1: "{{ groups['hdfs_nn'][0] | tosit.tdp.access_fqdn(hostvars) }}:8020"
...

Best regards

**Comment by rpignolet** [comment]
I understand the need to template `key` of `dict` inside `tdp_vars` but for me this not the good way to do it. We have a plugin `resolve` which merge and template `tdp_vars`. I think this should be possible to template `key` in the plugin code. If someone proves to me that this is not possible with the plugin, I will think about whether there is another solution.

**Comment by taabich** [comment]
Hi,
I'd be glad if  evaluating ` keys`  with ansible variable would work. but unfortunately it doesn't work. I tested several times before creating template-plus.
If you prove the pluging `resolve`  its works to evaluated {{ cluster_name | default('mycluster') }} in this case:
dfs.namenode.rpc-address.{{ cluster_name | default('mycluster') }}.nn1: "{{ groups['hdfs_nn'][0] | tosit.tdp.access_fqdn(hostvars) }}:8020"
I would be pleased  to delete template-plus.
Best regards

**Comment by leopaul36** [comment]
Created https://github.com/TOSIT-IO/tdp-collection/issues/469 for key name templating support in resolve``

**Comment by rpignolet** [comment]
@taabich  thank you for your contribution to TDP!
In a fully open-source approach, all Issues/Pull Requests are fairly examined for integration.
Please find below some points to address in order to validate your PR before merging to master branch:
  - `resolve` plugin now support template key with #482, use this feature and remove `template_plus`
  - consider remaining code review comments

Please note that without your feedback, this issue will be closed in a few days.
Thank you again for your contribution to TDP.

**Comment by leopaul36** [comment]
@taabich will you be working on this?

**Comment by taabich** [comment]
Hi! Sorry,
I'm busy at the moement, I would quickly take the resolution of the merge.
Best regards

**Comment by Nuttymoon** [comment]
Closing as this was fixed by #531

**Pull request closed without merging**